### PR TITLE
Improve native script handling

### PR DESCRIPTION
**Pull request by lehins**
# Description

* Remove redundant script hash verification. Script Hashes are not transmitted over the wire, they are computed upon deserialization. This means there is no point in recomputing and verifying those hashes as a predicate check in the rules
* Short circuit multisig and timelock verification for `RequireMOf` as soon as necessary number of scripts have been validated. Previously it was validating all scripts, regardless of how many was needed.
* Remove the whole Phased distinction, since it was not used nor is it needed for phase2 scripts. It was only necessary for native, phase1 scripts. That is why a simpler interface was added that deals with native scripts only. Addition of `NativeScript`, `getNativeScript` and `validateNativeScript`

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [x] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [x] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/input-output-hk/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

### CODE REVIEW

File: eras/allegra/impl/src/Cardano/Ledger/Allegra/Scripts.hs [hunk 1]
```diff
@@ -274,21 +273,24 @@ ltePosInfty (SJust i) j = i <= j
...
+    isValidMOf n SSeq.Empty = n <= 0
+    isValidMOf n (ts SSeq.:<| tss) =
+      n <= 0 || if go ts then isValidMOf (n - 1) tss else isValidMOf n tss
+    go = \case
+      RequireTimeStart lockStart -> lockStart `lteNegInfty` txStart
+      RequireTimeExpire lockExp -> txExp `ltePosInfty` lockExp
+      RequireSignature hash -> hash `Set.member` vhks
+      RequireAllOf xs -> all go xs
+      RequireAnyOf xs -> any go xs
+      RequireMOf m xs -> isValidMOf m xs
```
Comment (TimSheard) [inline eras/allegra/impl/src/Cardano/Ledger/Allegra/Scripts.hs:293]: could we call this something like 'validateTimeloxk'

File: eras/shelley/impl/src/Cardano/Ledger/Shelley/Scripts.hs [hunk 2]
```diff
@@ -191,3 +195,29 @@ eqMultiSigRaw t1 t2 = go (getMemoRawType t1) (getMemoRawType t2)
...
+    -- The important part of this validator is that it will stop as soon as it reaches the
+    -- required number of valid scripts
+    isValidMOf n [] = n <= 0
+    isValidMOf n (msig : msigs) =
+      n <= 0 || if go msig then isValidMOf (n - 1) msigs else isValidMOf n msigs
+    go = \case
+      RequireSignature hk -> Set.member hk vhks
+      RequireAllOf msigs -> all go msigs
+      RequireAnyOf msigs -> any go msigs
+      RequireMOf m msigs -> isValidMOf m msigs
```
Comment (TimSheard) [inline eras/shelley/impl/src/Cardano/Ledger/Shelley/Scripts.hs:217]: How about 'evalMultiSig'  instead of go

**Review by TimSheard — APPROVED** [review]
Seems like a better approach.  using 'go' seems a bit too obtuse though.

File: eras/shelley/impl/src/Cardano/Ledger/Shelley/Scripts.hs (hunk 2 above)
Comment (lehins) [reply eras/shelley/impl/src/Cardano/Ledger/Shelley/Scripts.hs:217]: `evalMultiSig` is the name of the function. :smile: 

Are you talking about the local `isValidMOf`? This one only works for `MOf`, hence the name `isValidMOf`

File: eras/allegra/impl/src/Cardano/Ledger/Allegra/Scripts.hs (hunk 1 above)
Comment (lehins) [reply eras/allegra/impl/src/Cardano/Ledger/Allegra/Scripts.hs:293]: Same here. `evalTimelock` is the name of the function. `isValidMOf` is only applicable to the `RequireMOf`

Comment (lehins) [reply eras/allegra/impl/src/Cardano/Ledger/Allegra/Scripts.hs:293]: `validateTimelock` is the top level function right here that calls `evalTimelock`:

https://github.com/input-output-hk/cardano-ledger/blob/71bf6ae648fdead3bf37d5ba69d46f9a35d29dc2/eras/allegra/impl/src/Cardano/Ledger/Allegra/Tx.hs#L72-L77

**Pull request merged**